src/libraries/Common/src/System/Security/Cryptography/Oids.Shared.cs

```diff
@@ -67,5 +71,79 @@ private static Oid InitializeOid(string oidValue)
             return oid;
         }
 
+        internal static Oid GetSharedOrNewOid(ref AsnValueReader asnValueReader)
+        {
+            Oid? ret = GetSharedOrNullOid(ref asnValueReader);
+
+            if (ret is not null)
+            {
+                return ret;
+            }
+
+            string oidValue = asnValueReader.ReadObjectIdentifier();
+            return new Oid(oidValue, null);
+        }
+
+        internal static Oid? GetSharedOrNullOid(ref AsnValueReader asnValueReader, Asn1Tag? expectedTag = null)
+        {
+#if NET
+            Asn1Tag tag = asnValueReader.PeekTag();
+
+            // This isn't a valid OID, so return null and let whatever's going to happen happen.
+            if (tag.IsConstructed)
+            {
+                return null;
+            }
+
+            Asn1Tag expected = expectedTag.GetValueOrDefault(Asn1Tag.ObjectIdentifier);
+
+            Debug.Assert(
+                expected.TagClass != TagClass.Universal ||
+                expected.TagValue == (int)UniversalTagNumber.ObjectIdentifier,
+                $"{nameof(GetSharedOrNullOid)} was called with the wrong Universal class tag: {expectedTag}");
+
+            // Not the tag we're expecting, so don't match.
+            if (!tag.HasSameClassAndValue(expected))
+            {
+                return null;
+            }
+
+            ReadOnlySpan<byte> contentBytes = asnValueReader.PeekContentBytes();
+
+            Oid? ret = contentBytes switch
+            {
+                [0x55, 0x04, 0x03] => CommonNameOid,
+                _ => null,
+            };
+
+            if (ret is not null)
+            {
+                // Move to the next item.
+                asnValueReader.ReadEncodedValue();
+            }
+
+            return ret;
+#else
+            // Cannot use list patterns on older TFMs.
+            return null;
+#endif
```

Comment:
Why would we not just implement the few lines:
```C#
            Oid? ret = contentBytes switch
            {
                [0x55, 0x04, 0x03] => CommonNameOid,
                _ => null,
            };
```
as:
```C#
Oid? ret = contentBytes.Length == 3 && contentBytes[0] == 0x55 && contentBytes[1] == 0x44 && contentBytes[2] == 0x03 ? CommonNameOid : null;
```
rather than not implementing this method at all downlevel?

Reply:
I suppose the idea is to list the ~6-8 common Oids that appear in the RDNs.

Reply:
> I suppose the idea is to list the ~6-8 common Oids that appear in the RDNs.

Sorry, I'm not following.  The comment suggests that the whole method implementation is ifdef'd out purely because of the inability to use list patterns, which appear to only be used in one place that's easily replaceable with something that does work.

Reply:
https://github.com/dotnet/runtime/pull/72235#discussion_r921721562

The idea is to avoid allocations for the common values I think.

Reply:
The intention is that several more cases are going to be added:

```C#
Oid? ret = contentBytes switch
            {
                [0x55, 0x04, 0x03] => CommonNameOid,
                [0x55, 0x04, 0x06] => CountryNameOid,
                [0x55, 0x04, 0x07] => LocalityOid,
                // etc
                _ => null,
            };
```

Reply:
This file is used in two main places: System.Security.Cryptography.dll (shared runtime, NetCoreAppCurrent) and System.Security.Cryptography.Pkcs (OOB, includes a netstandard2.0 target, a ns2.1 target, some netcoreapp targets, and netfx).

Given how broad a TFM set it builds with, I doubt anyone is even running the netstandard2.0 build, but it makes the package complete for the ns2.0 ref.  The netfx build is mostly a facade to GAC:System.Security.dll, but there are probably a few common OIDs that could get added to the `#else` block as we start moving things from `AsnReader.ReadObjectIdentifier()` to a "your bytes look familiar, here's the answer" approach.

In the medium to long term, I'd expect most (if not all) of the shared OIDs in this file to be represented in the pattern-match, most of which aren't relevant to the parts of the Pkcs library that aren't typeforwards on .NET Framework.  So it's really a question of overall best-fit:

* Use the list pattern for newer targets, ignore the old ones as low usage.
* Use the list pattern for newer targets, add some tracing in the `#else` and add ones that get hit running PKCS7 tests on netfx.
* Use the expanded form everywhere.
* Keep doing what we've been doing (full data interpretation and building new System.String instances for every call to ReadObjectIdentifier)

I started with expanded, then @vcsjones showed me the list pattern, and I thought it read better, so I'm at 1 with an easy future pivot to 2.  But I'm definitely willing to be convinced to go with 3 or 4.

Reply:
If the method is just an optimization and we don't think that optimization matters in the few to none cases where it'll be used downlevel, seems fine to just leave the code as-is.  It just seemed that there was this long method, seemingly important, and entirely commented out if a language feature wasn't available to just save a few lines of code.

Reply:
```diff
 #else
-            // Cannot use list patterns on older TFMs.
+            // The list pattern isn't available in System.Security.Cryptography.Pkcs for the
+            // netstandard2.0 or netfx builds.  Any OIDs that it's important to optimize in
+            // those contexts can be matched on here, but using a longer form of matching.
+
             return null;
 #endif
```

Better?

Reply:
Thanks :)